libcontainer/nsenter/cloned_binary.c

```diff
@@ -86,15 +87,34 @@ static void *must_realloc(void *ptr, size_t size)
 /*
  * Verify whether we are currently in a self-cloned program (namely, is
  * /proc/self/exe a memfd). F_GET_SEALS will only succeed for memfds (or rather
- * for shmem files), and we want to be sure it's actually sealed.
+ * for shmem files), and we want to be sure it's actually sealed. In case the
+ * the program is on a read-only rootfs, do not clone.
  */
 static int is_self_cloned(void)
 {
 	int fd, ret, is_cloned = 0;
 	struct stat statbuf = {};
 	struct statfs fsbuf = {};
+	struct statvfs stat;
+	char abs_path[1024];
+
+	memset(&stat, 0, sizeof(struct statvfs));
+	const char *path = "/proc/self/exe";
+
+	ssize_t len = readlink(path, abs_path, sizeof(abs_path)-1);
+	if (len != -1) {
+		/* readlink does not provide the path with null terminated character. */
+		abs_path[len] = '\0';
+
+		if (statvfs(abs_path, &stat) == 0) {
```

Comment:
When the actual length of the path is over 1024, this is not reliable

Reply:
I was not sure of having an absolute path corresponding to more than 1024 characters. But, yes using `PATH_MAX` for the array size will be a better approach as described by cyphar@.